website/blog/2019-02-22-shadow-dom-support.md

```diff
@@ -0,0 +1,186 @@
+
+ The component object pattern attempts to reduce that repetition and move the component's api into an object of it's own. We know that in order to interact with an element's shadow DOM, we first need the host element. Using a base class for your component objects makes this pretty straightforward. Here's a bare-bones component base class that takes the `host` element in it's contructor and unrolls that element's queries up to the browser object, so it can be reused in many page objects (or other component objects), without having to know anything about the page itself:
+
+```javascript
+class Component {
+
+  constructor(host) {
+    const selectors = [];
+    // Crawl back to the browser object, and cache all selectors
+    while (host.elementId && host.parent) {
+      selectors.push(host.selector);
+      host = host.parent;
+    }
+    selectors.reverse();
+    this.selectors_ = selectors;
+  }
+
+  get host() {
+    // Beginning with the browser object, reselect each element
+    return this.selectors_.reduce((element, selector) => element.$(selector), browser);
+  }
+}
+
+module.exports = Component;
+```
+
+We can then write a subclass for our app-login component:
+
+```javascript
+const Component = require('./component');
+
+class Login extends Component {
+
+  get usernameInput() {
+    return this.host.shadow$('input #username');
+  }
+
+  get passwordInput() {
+    return this.host.shadow$('input[type=password]');
+  }
+
+  get submitButton() {
+    return this.login.shadow$('button[type=submit]');
+  }
+
+  login(username, password) {
+    this.usernameInput.setValue(username);
+    this.passwordInput.setValue(password);
+    this.submitButton.click();
+  }
+}
+
+module.exports = Login;
+```
+
+Finally, we can use the component object inside our login page object:
+
+```javascript
+const Login = require('./components/login');
+
+class LoginPage {
+
+  open() {
+    browser.url('/login');
+  }
+
+  get app() {
+    return browser.$('my-app');
+  }
+
+  get loginComponent() {
+    // return a new instance of our login component object
+    return new Login(this.app.$('app-login'));
+  }
+
+}
+```
+This component object can now be used in tests for any page or section of your app that uses an app-login web component, without having to know about how that component is structured. If you later decide to change the internal structure of the web component, you only need to update the component object.
+
+
+## Future
+Currently Selenium does not provide native support for shadow DOM, and the w3c webdriver spec writers are still up in the air as to how it should be done. Once the spec is finalized, WebdriverIO will implement the spec. There's a decent chance that the `shadow` commands will change under the hood, but I'm pretty confident that they're usage will be the same as it is today, and that test code that uses them will need little to no refactoring.
```

Comment:
```suggestion
Currently the [WebDriver protocol](https://w3c.github.io/webdriver/) does not provide native support for shadow DOM, but there has been [progress](https://github.com/w3c/webdriver/issues/350) made for it. Once the spec is finalized, WebdriverIO will implement the spec. There's a decent chance that the `shadow` commands will change under the hood, but I'm pretty confident that they're usage will be the same as it is today, and that test code that uses them will need little to no refactoring.
```

There is no reason to mention Selenium here. Also we have been making progress on the shadow DOM proposal at last TPAC meeting, so let's link the issue there so people can follow up on it.